### PR TITLE
fix(modal): prevent body content shifting when vertical scrollbar

### LINKED ISSUE #3714: Open modal is shifting body content to the right
When the page has a vertical scroll and opens a modal window, the vertical scroll on the page hiding. Because of this, the content moves to the right. In the Bootstrap this problem is solved by adding right padding to the body when scroll hiding.

http://plnkr.co/edit/ynWJUbnryVkSHdEjFRMw?p=preview

### PR DESCRIPTION
**Pull request by kunal999**
Prevents page body content shifting to the right when modal is open (when appendTo is 'body')

Closes #3714

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -15,6 +15,7 @@
     "angular": "^1.4.4",
     "angular-mocks": "^1.4.4",
     "angular-sanitize": "^1.4.4",
+    "bootstrap": "^3.3.5",
```
Comment (wesleycho) [inline package.json]: Undo this change.

File: karma.conf.js [hunk 2]
```diff
@@ -19,6 +19,7 @@ module.exports = function(config) {
       'node_modules/angular/angular.js',
       'node_modules/angular-mocks/angular-mocks.js',
       'node_modules/angular-sanitize/angular-sanitize.js',
+      'node_modules/bootstrap/dist/css/bootstrap.css',
```
Comment (wesleycho) [inline karma.conf.js]: Does this affect the test?

File: src/modal/modal.js [hunk 3]
```diff
@@ -463,6 +521,25 @@ angular.module('ui.bootstrap.modal', ['ui.bootstrap.stackedMap'])
...
+          // set bodyOverflowing property
+          modalDomEl.bodyOverflowing = true;
+
+          // set right padding to body
+          if(openedWindows.length() === 1){
+            if(!$$scrollbarHelper.scrollbarWidth){
+              $$scrollbarHelper.measureScrollbar();
+            }
+
+            var body = $document.find('body').eq(0);
```
Comment (wesleycho) [inline src/modal/modal.js]: No need to do `.eq(0)` here, should only be one body tag.

File: karma.conf.js (hunk 2 above)
Comment (kunal999) [inline karma.conf.js]: Yes. Without bootstrap.css, the tests always pass. The tests rely on the modal related classes in bootstrap css which remove the scroll bar from the body when the modal is opened.

File: src/modal/modal.js (hunk 3 above)
Comment (kunal999) [inline src/modal/modal.js]: Removed the unnecessary `.eq(0)`

File: src/modal/modal.js [hunk 4]
```diff
@@ -129,6 +151,33 @@ angular.module('ui.bootstrap.modal', ['ui.bootstrap.stackedMap'])
...
+          // check if modal is overflowing
+          var modalIsOverflowing = element[0].scrollHeight > document.documentElement.clientHeight;
+          
+          if(!$$scrollbarHelper.scrollbarWidth){
+            $$scrollbarHelper.measureScrollbar();
+          }
+
+          element.css({
+            'padding-left': (!bodyIsOverflowing && modalIsOverflowing ? $$scrollbarHelper.scrollbarWidth : '') + 'px',
+            'padding-right': (bodyIsOverflowing && !modalIsOverflowing ? $$scrollbarHelper.scrollbarWidth : '') + 'px'
```
Comment (wesleycho) [inline src/modal/modal.js]: This means that the object could potentially be setting
```js
element.css({
  'padding-left': 'px',
  'padding-right': 'px'
});
```
It should be setting 0 at worse.

File: src/modal/modal.js [hunk 5]
```diff
@@ -129,6 +151,33 @@ angular.module('ui.bootstrap.modal', ['ui.bootstrap.stackedMap'])
         element.addClass(attrs.windowTopClass || '');
         scope.size = attrs.size;
 
+        // get topmost modal object
+        var bodyIsOverflowing;
+        var modal = $modalStack.getTop();
+        
+        // only when modal is attached to body
+        if(modal && modal.value && modal.value.appendTo === 'body'){
```
Comment (wesleycho) [inline src/modal/modal.js]: Shouldn't `modal` exist if the `uib-modal-window` directive is used? Also shouldn't `modal.value' & `modal.value.appendTo` exist at all times?

File: src/modal/modal.js (hunk 4 above)
Comment (kunal999) [inline src/modal/modal.js]: Fixed to set the `padding-left` and `padding-right` only when required

File: src/modal/modal.js (hunk 5 above)
Comment (kunal999) [inline src/modal/modal.js]: Removed unnecessary null checks

File: src/modal/modal.js [hunk 6]
```diff
@@ -106,9 +106,31 @@ angular.module('ui.bootstrap.modal', ['ui.bootstrap.stackedMap'])
...
+    function measureScrollbar () { // thx walsh
+      var scrollDiv = document.createElement('div');
+      scrollDiv.className = 'modal-scrollbar-measure';
+      document.body.appendChild(scrollDiv);
+      var width = scrollDiv.offsetWidth - scrollDiv.clientWidth;
+      document.body.removeChild(scrollDiv);
+      this.scrollbarWidth = width;
+    }
+  })
+
```
Comment (RobJacobs) [inline src/modal/modal.js]: I'd like to see this in the $position service as scrollbarWidth(), we might need to use it for auto/smart positioning. Would be nice to cache the result to avoid manipulating the DOM every time we need this value as it should not change.

Comment (wesleycho) [inline src/modal/modal.js]: I agree - we have a related issue in #4317.

**Comment by wesleycho** [comment]
Can you squash then rebase this, and then change to use the `$uibPosition.scrollbarWidth` method for calculating the scrollbar width?

**Comment by wesleycho** [comment]
Also, please fix the code style to match the rest of the file.

File: src/modal/modal.js [hunk 7]
```diff
@@ -202,6 +224,34 @@ angular.module('ui.bootstrap.modal', ['ui.bootstrap.stackedMap'])
...
+            if(modal.value.appendTo === 'body'){
+              
+              // check bodyOverflowing property that was set when opening modal
+              if(modal.value.modalDomEl.bodyOverflowing){
+                bodyIsOverflowing = true;
+              }
+
+              // start - from adjustDialog method of modal.js of bootstrap
+              // check if modal is overflowing
+              modalIsOverflowing = element[0].scrollHeight > document.documentElement.clientHeight;
```
Comment (wesleycho) [inline src/modal/modal.js]: Should use `$document[0].clientHeight`

File: src/modal/modal.js [hunk 8]
```diff
@@ -463,6 +522,25 @@ angular.module('ui.bootstrap.modal', ['ui.bootstrap.stackedMap'])
         appendToElement.append(modalDomEl);
         appendToElement.addClass(modalBodyClass);
 
+        // if modal is attached to body, then check if body is overflowing and set bodyOverflowing property on modal object and set right padding equal to scrollbarWidth
+        if (modal.appendTo === 'body' && document.body.scrollHeight > document.documentElement.clientHeight){
```
Comment (wesleycho) [inline src/modal/modal.js]: Ditto

File: src/modal/modal.js [hunk 9]
```diff
@@ -463,6 +522,25 @@ angular.module('ui.bootstrap.modal', ['ui.bootstrap.stackedMap'])
...
+
+          // set right padding to body
+          if(openedWindows.length() === 1){
+            if(!$$scrollbarHelper.scrollbarWidth){
+              $$scrollbarHelper.measureScrollbar();
+            }
+
+            var body = $document.find('body');
+            // set originalBodyRightPadding property to reset it when the last modal closes
+            modalDomEl.originalBodyRightPadding = body.css('padding-right');
```
Comment (wesleycho) [inline src/modal/modal.js]: That is an unnecessarily verbose key name - please choose a better name, such as `bodyRightPadding`.

File: src/modal/modal.js [hunk 10]
```diff
@@ -291,6 +343,13 @@ angular.module('ui.bootstrap.modal', ['ui.bootstrap.stackedMap'])
           openedClasses.remove(modalBodyClass, modalInstance);
           appendToElement.toggleClass(modalBodyClass, openedClasses.hasKey(modalBodyClass));
           toggleTopWindowClass(true);
+
+          // reset any right padding set to body when opening the modal
+          // make sure no other modal is open before resetting
+          if(modalWindow.appendTo === 'body' && openedWindows.length() === 0){
+            var body = $document.find('body');
+            body.css('padding-right', modalWindow.modalDomEl.originalBodyRightPadding);
```
Comment (wesleycho) [inline src/modal/modal.js]: This seems way too opinionated - this will make it near impossible for a user to override this, and is not a good solution.

File: src/modal/modal.js (hunk 6 above)
Comment (RobJacobs) [inline src/modal/modal.js]: $position service now has a scrollbarWidth function.

**Comment by wesleycho** [comment]
Going to close this due to the issues raised in the code review & due to the merge conflict - feel free to open a new PR fixing this along with a cleaner implementation.
